### PR TITLE
Implement python binding generator for embedded_messages

### PR DESCRIPTION
**Pull request by daw1012345**
Everything is described in the readme. The implementation is versatile - if the protocol evolves in the future, bindings should very easy to regenerate to take the changes into account. 

This is achieved by not hardcoding the cffi cdefs, but by doing some simple C parsing with pycparser to turn type/function definitions into declarations (as required by cffi).

### CODE REVIEW

File: python_bindings/generate.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+import os
+import argparse
+from collections import OrderedDict
+
+try:
+    # For generating bindings
+    import cffi 
+except (ImportError):
```
Comment (Yuhanun) [inline python_bindings/generate.py]:
```suggestion
except ImportError:
```

File: python_bindings/generate.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    print("[!] `CFFI` is missing.")
+    exit(1)
+
+# pycparser is a dependency of cffi, so must be available here
+from pycparser import parse_file, c_generator, c_ast 
+
+# The C pre-processor expands the fake stdlib headers we provide. If this is compiled by cffi, there will be weird type size issues.
+def kill_unwanted_typedefs(tree):
+    for x, i in tree.children():
```
Comment (Yuhanun) [inline python_bindings/generate.py]:
```suggestion
    for _, i in tree.children():
```

File: python_bindings/generate.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        # We only need to look at Typedefs
+        if type(i) is c_ast.Typedef:
+            # If the type is too simple or if it's a struct with no implementation
+            if type(i.type.type) is c_ast.IdentifierType or (type(i.type.type) is c_ast.Struct and not i.type.type.decls):
+                tree.ext.remove(i)
```
Comment (Yuhanun) [inline python_bindings/generate.py]: Flip theses 2 conditions around, if not type(i) is ... continue

File: python_bindings/generate.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    exit(1)
+
+# pycparser is a dependency of cffi, so must be available here
+from pycparser import parse_file, c_generator, c_ast 
+
+# The C pre-processor expands the fake stdlib headers we provide. If this is compiled by cffi, there will be weird type size issues.
+def kill_unwanted_typedefs(tree):
+    for x, i in tree.children():
+        # We only need to look at Typedefs
+        if type(i) is c_ast.Typedef:
```
Comment (Yuhanun) [inline python_bindings/generate.py]: instanceof()

File: python_bindings/generate.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            # If the type is too simple or if it's a struct with no implementation
+            if type(i.type.type) is c_ast.IdentifierType or (type(i.type.type) is c_ast.Struct and not i.type.type.decls):
```
Comment (Yuhanun) [inline python_bindings/generate.py]: Same for these and the following ones

File: python_bindings/generate.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            # Some edge cases - a typedef with pointers
+            elif type(i.type) is c_ast.PtrDecl:
+                if type(i.type.type.type) is c_ast.IdentifierType:
+                    tree.ext.remove(i)
+
+
+def expose_decl_in_bindings(files: list, ffi: cffi.FFI, c_compiler: str):
+    # Create a temporary file as workaround for issues with redeclaring types
+    with open('.tmp_preprocess.h', 'w') as tmp_file:
+        tmp_file.write(format_include_for_compiler(files))
```
Comment (Yuhanun) [inline python_bindings/generate.py]: https://docs.python.org/3/library/io.html

File: python_bindings/.gitignore [hunk 7]
```diff
@@ -0,0 +1,3 @@
+pycparser/
+.tmp_preprocess.h
+.DS_Store
```
Comment (Yuhanun) [inline python_bindings/.gitignore:3]: ah, macOS?

File: python_bindings/generate.py (hunk 6 above)
Comment (Yuhanun) [reply python_bindings/generate.py]: Memory files

File: python_bindings/README.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Python bindings generator for roboteam_embedded_messages
+
+## Usage
+
+1. (Optional) Enter a python virtual env
+2. Run bash.sh to setup the enviornemnt
```
Comment (emielsteerneman) [inline python_bindings/README.md]: Should this be setup.sh?

File: python_bindings/generate.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        final += f"#include \"{os.path.abspath(file)}\"\n"
+    return final
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Generates python bindings for given files')
+    parser.add_argument("--includes", required=True, nargs='+', help="C header files to make bindings for")
+    parser.add_argument("--output", nargs=1, type=str, help="Output location of the python bindings", default=".")
+    parser.add_argument("--name", nargs=1, type=str, help="Name of the generated python module", default="_roboteam_embedded_messages")
+    parser.add_argument("--lib", nargs='+', help="Additional libraries to link with", default=[])
+    parser.add_argument("--compiler", nargs=1, type=str, help="C(++) compiler to use to preprocess files", default="gcc")
```
Comment (emielsteerneman) [inline python_bindings/generate.py]: Remove `nargs=1` when `type=str`. `nargs=1` turns the `str` into `[str]`, which crashes cffi. 
tested with `python3 generate.py --includes ../include/RobotCommand.h --output "./folder"`

File: python_bindings/generate.py (hunk 6 above)
Comment (daw1012345) [reply python_bindings/generate.py]: I needed an actual file on disk, so I instead used the python tempfile module. Is that satisfactory?

File: python_bindings/.gitignore (hunk 7 above)
Comment (daw1012345) [reply python_bindings/.gitignore:3]: Yep, and .DS_Store can leak info...

File: python_bindings/generate.py (hunk 6 above)
Comment (Yuhanun) [reply python_bindings/generate.py]: Yeah sure.

**Review by Yuhanun — APPROVED** [review]
LGTM

**Review by emielsteerneman — CHANGES_REQUESTED** [review]

File: python_bindings/README.md (hunk 8 above)
Comment (daw1012345) [reply python_bindings/README.md]: Yep, it was a typo. Fixed now

File: python_bindings/generate.py (hunk 9 above)
Comment (daw1012345) [reply python_bindings/generate.py]: Didn't know about this behaviour, removed the redundant nargs.

**Review by emielsteerneman — APPROVED** [review]

**Pull request merged**